### PR TITLE
themes/pzq: Add pzq themes inspired by ys-theme from oh-my-zsh and rana

### PR DESCRIPTION
**Pull request by ZQPei**
* Add pzq theme, which is inspired on ys-theme from oh-my-zsh and rana theme
* The pzq theme solves the problem of misplaced commands in rena

### CODE REVIEW

File: themes/pzq/pzq.theme.sh [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+is_vim_shell() {
+  if [ ! -z "$VIMRUNTIME" ];
+  then
+    echo "${D_INTERMEDIATE_COLOR}on ${D_VIMSHELL_COLOR}\
+vim shell${D_DEFAULT_COLOR} "
+  fi
+}
+
+mitsuhikos_lastcommandfailed() {
+  code=$?
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:47]:
```suggestion
  local code=$?
```

File: themes/pzq/pzq.theme.sh [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+		echo -e "${D_GIT_DEFAULT_COLOR}on ${D_GIT_BRANCH_COLOR}${branchName} ";
+	else
+		return;
+	fi;
+}
+
+limited_pwd() {
+  # Max length of PWD to display
+  MAX_PWD_LENGTH=20
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:93]: Do we want to expose the value to users? If not, we can make it local.

```suggestion
  local MAX_PWD_LENGTH=20
```

File: themes/pzq/pzq.theme.sh [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+  # Replace $HOME with ~ if possible 
+  RELATIVE_PWD=${PWD/#$HOME/\~}
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:96]:
```suggestion
  local RELATIVE_PWD=${PWD/#$HOME/\~}
```

File: themes/pzq/pzq.theme.sh [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+  local offset=$((${#RELATIVE_PWD}-$MAX_PWD_LENGTH))
+
+  if [ $offset -gt "0" ]
+  then
+      local truncated_symbol="..."
+      TRUNCATED_PWD=${RELATIVE_PWD:$offset:$MAX_PWD_LENGTH}
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:103]:
```suggestion
      local TRUNCATED_PWD=${RELATIVE_PWD:$offset:$MAX_PWD_LENGTH}
```

File: themes/pzq/pzq.theme.sh [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+      echo -e "${truncated_symbol}/${TRUNCATED_PWD#*/}"
+  else
+      echo -e "${RELATIVE_PWD}"
+  fi
+}
+
+prompt_python() {
+  # for virtualenv
+  if [[ $VIRTUAL_ENV ]]; then
+    virtualenv=`basename "$VIRTUAL_ENV"`
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:113]:
```suggestion
    local virtualenv=$(basename "$VIRTUAL_ENV")
```

Prefer `$(...)` to ``` `...` ``` which is deprecated by POSIX.

File: themes/pzq/pzq.theme.sh [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    VIRTUALENV_THEME_PROMPT_PREFIX='('
+    VIRTUALENV_THEME_PROMPT_SUFFIX=')'
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:115]:
```suggestion
    local VIRTUALENV_THEME_PROMPT_PREFIX='('
    local VIRTUALENV_THEME_PROMPT_SUFFIX=')'
```

File: themes/pzq/pzq.theme.sh [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+  local LAST_COMMAND_FAILED=$(mitsuhikos_lastcommandfailed)
+  local SAVE_CURSOR='\033[s'
+  local RESTORE_CURSOR='\033[u'
+  local MOVE_CURSOR_RIGHTMOST='\033[500C'
+  local MOVE_CURSOR_5_LEFT='\033[5D'
+  local THEME_CLOCK_FORMAT="%H:%M:%S %y-%m-%d"
+  # Replace $HOME with ~ if possible 
+  local RELATIVE_PWD=${PWD/#$HOME/\~}
+
+  if [ $(uname) = "Linux" ];
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:138]:
```suggestion
  if [ "$OSTYPE" = "linux-gnu" ];
```

File: themes/pzq/pzq.theme.sh [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    virtualenv=`basename "$VIRTUAL_ENV"`
+    VIRTUALENV_THEME_PROMPT_PREFIX='('
+    VIRTUALENV_THEME_PROMPT_SUFFIX=')'
+    echo -e "${VIRTUALENV_THEME_PROMPT_PREFIX}$virtualenv${VIRTUALENV_THEME_PROMPT_SUFFIX}"
+  fi
+
+  # for condaenv
+  if [[ $CONDA_DEFAULT_ENV ]]; then
+    CONDAENV_THEME_PROMPT_PREFIX='('
+    CONDAENV_THEME_PROMPT_SUFFIX=')'
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:122]:
```suggestion
    local CONDAENV_THEME_PROMPT_PREFIX='('
    local CONDAENV_THEME_PROMPT_SUFFIX=')'
```

File: themes/pzq/pzq.theme.sh [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    echo -e "${CONDAENV_THEME_PROMPT_PREFIX}${CONDA_DEFAULT_ENV}${CONDAENV_THEME_PROMPT_SUFFIX}"
+  fi
+}
+
+# -------------------------------------------------------------- PROMPT OUTPUT
+prompt() {
+  local LAST_COMMAND_FAILED=$(mitsuhikos_lastcommandfailed)
+  local SAVE_CURSOR='\033[s'
+  local RESTORE_CURSOR='\033[u'
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:131]: Modern terminals may not support these SCO sequences. Most terminals support `\e7` and `\e8`, respectively. Also, by using `\[...\]`, we need to let Bash know that the cursor movements between these two values are canceled.

```suggestion
  local SAVE_CURSOR='\[\e7'
  local RESTORE_CURSOR='\e8\]'
```

Or, if you want to make them work in a wider range of terminals, you could obtain them through `tput sc` and `tput rc`, respectively.

File: themes/pzq/pzq.theme.sh [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+  local MOVE_CURSOR_RIGHTMOST='\033[500C'
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:132]: 500 might not be sufficient for this 4K era...

```suggestion
  local MOVE_CURSOR_RIGHTMOST='\033['${COLUMNS:-9999}'C'
```

File: themes/pzq/pzq.theme.sh [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+D_CHANGES_COLOR="${white}"
+D_TIME_COLOR="${white}"
+D_DOLLOR_COLOR="${bold_red}"
+D_CMDFAIL_COLOR="${red}"
+D_VIMSHELL_COLOR="${cyan}"
+
+# ------------------------------------------------------------------ FUNCTIONS
+case $TERM in
+  xterm*)
+      TITLEBAR="\033]0;\w\007"
```
Comment (akinomyoga) [inline themes/pzq/pzq.theme.sh:31]:
```suggestion
      TITLEBAR="\[\033]0;\w\e\\\]"
```

- OSC terminated by BEL is deprecated. It is recommended to use the ANSI form terminated by ST (`\e\\`).
- We need to enclose it with `\[...\]` to let Bash know that the cursor will not be moved by the sequence.

**Review by akinomyoga — CHANGES_REQUESTED** [review]

**Comment by akinomyoga** [comment]
Please also add an image of the theme and update [themes/THEMES.md](https://github.com/ohmybash/oh-my-bash/blob/master/themes/THEMES.md)!

**Comment by akinomyoga** [comment]
I'm thinking of taking over this PR. If you still would like to work on it by yourself, please let me know.

**Review by akinomyoga — APPROVED** [review]

**Comment by akinomyoga** [comment]
I have rebased the PR on top of the latest master, applied suggested changes, and added other fixes. I have also added an image of the theme in `themes/THEMES.md`. I'll keep this PR for about one week and am going to merge it if there is no further comments. @nntoan If you would like to review the PR, please let me know, in which case I'll wait for your review before merging.

**Comment by ZQPei** [comment]
LGTM

**Comment by akinomyoga** [comment]
@ZQPei Thank you for checking the changes to the PR!

**Pull request merged**